### PR TITLE
feat: http2 support for http matcher

### LINKED ISSUE #51: [Question]: layer4 module seems cannot match http/2 or gRPC's header while caddy server does ?
The layer4 module seems cannot reconizge grpc/http2 connections with configuration below.
There's no upstream record in logs :
```
        { "handle":[{"handler":"tls"}]},
        { "handle":[
            {"handler":"proxy","upstreams":[{"dial":["127.0.0.1:8003"]}]}
          ],
          "match":[{"http":[{"header":{"Content-Type":["application/grpc*"]}}]}
          ]
        { "handle":[
            {"handler":"proxy","upstreams":[{"dial":["127.0.0.1:8003"]}]}
          ],
          "match":[{"http":[{"path":["/grpc"]}]}
          ]
        },
```
Both matchers does not work.

I only find a way to upstream gRPC now is using layer4 to proxy all connections (incl gRPC) into caddy server and then use reverse_proxy by match Content-Type or path.
example:
```
          { "handle":[{
              "handler":"reverse_proxy",
              "flush_interval":-1,
              "upstreams":[{"dial":gRPC Server"}],
              "transport":{"protocol":"http","versions":["h2c"]}}],
            "match":[{"header":{"Content-Type":["application/grpc*"]},"path":["/grpc"]}]},
```
So what's the difference between caddy server and layer4 in gRPC's header matcher ?
May I upstream gRPC connections by using layer4 module ?

### PR DESCRIPTION
**Pull request by ydylla**
Closes #51

### CODE REVIEW

**Comment by ydylla** [comment]
@mholt I have added some http matcher tests. You can review it now.  
I am also using it since yesterday with real http2 traffic from browsers and until now I did not see any errors in the logs. So I am pretty confident I did not miss something important 😄

**Comment by ydylla** [comment]
I just noticed I forgot to test the "protocol" matcher and it also does not work. Requests parsed by layer4 never have a filled TLS field, which the caddy `MatchProtocol` matcher uses. But I think the http matcher could fake one and fill it with information from the ClientHelloInfo that is saved at `cx` when the tls handler was executed before. Or at least with the server name when only the tls matcher was executed before.  
Is this a reasonable idea @mholt?

**Comment by mholt** [comment]
Ah yeah, that sounds good to me. Let's try it out!

**Comment by ydylla** [comment]
I will give it a try tomorrow and push it here.

**Comment by ydylla** [comment]
@mholt Instead of faking the tls connection state I decided to actually use the real one. The tls handler now saves them to the layer4 connection similar to the client hellos.  

Now it's possible to use a config with a single server that listens to port 80 & 443, accepts http & https traffic on both and also redirects http traffic to use https. The single server config block is not really necessary, but it's fun 😄

**Comment by ydylla** [comment]
Hi @mholt,
did you find time to review this?  
I am already using it and it works fine, so I think it can be merged.

**Comment by ydylla** [comment]
Hi @mholt,
did you find time to review this?  
I am already using it and it works fine, so I think it can be merged.

**Comment by mholt** [comment]
Eek, I did not yet. I was busy with the release of Caddy 2.5. sorry. I will try to look at this weekend or early next week. I think this is really cool!

File: modules/l4http/httpmatcher.go [hunk 1]
```diff
@@ -96,6 +113,80 @@ func (m MatchHTTP) Match(cx *layer4.Connection) (bool, error) {
...
+	}
+
+	// According to http.Request.URL docs it only contains the value of RequestURI (so path only),
+	// but we can fill in more information
+	req.URL, err = url.Parse(fmt.Sprintf("%s://%s%s", scheme, authority, path))
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (mholt) [inline modules/l4http/httpmatcher.go:187]:
```suggestion
	req.URL, err = url.Parse(fmt.Sprintf("%s://%s%s", scheme, authority, path))
	return err
```

**Review by mholt — APPROVED** [review]
Very nice -- LGTM. Just one nit, but I think this is a quality change. A bit unfortunate we have to do so much of the parsing work ourselves, but oh well. Nicely done! Let's try it out.

**Comment by ydylla** [comment]
Thanks for the review.  
I actually prefer the more verbose version and in this particular case it was because I commented it out and in multiple times 😄 
Should I click on merge or do you want to do it?

**Comment by mholt** [comment]
@ydylla You do the honors!

**Pull request merged**